xds/type/matcher/v3/string.proto

```diff
@@ -52,6 +52,23 @@ message StringMatcher {
     //
     // * *abc* matches the value *xyz.abc.def*
     string contains = 7 [(validate.rules).string = {min_len: 1}];
+
+    // The input string must have the pattern specified here.
+    // The space between slashes acts as variable/capture group
+    // but does not have any predefined validation.
+    // The variables names are user defined.
+    //
+    // Empty pattern match is not allowed.
+    //
+    // Note: There is overlap between regex and pattern.
+    // Pattern serves as a better user experience through simplified matching syntax.
+    //
+    // Examples:
+    //
+    // abc/{lang}/{county} matches the value abc/eng/us
+    // Note format of: abs/<user-defined-var>/<user-defined-var>
+    // {lang} == "eng" and {country} == "us"
+    string pattern = 8 [(validate.rules).string = {min_len: 1}];
```

Comment:
Sorry I think I'm missing something.
Is the idea to match a string given some pattern, and assign the values to their variables (e.g., lang will be "eng"), or are there known categories and known possible values for these categories, and a match will occur if the used values are in the categories.

Reply:
The will be matched and assigned to variables. No know categories.

Reply:
How would the string "abcdef" match against pattern="abc{first}{second}", as the possible assignments are possible:
first="de", second="f"
first="d", second="ef"
?

Reply:
This would not be a valid config. Only items between "/"s are used as vars.

Reply:
This is an interesting use-case, and I'm not sure if it should be specified as part of the StringMatcher (the StringMatcher just matches according to some pattern, and doesn't assign values to external variables).

@snowp was this kind of use-case ever considered?

Reply:
The variable extraction will be done elsewhere. Let's just keep the matcher a matcher

Reply:
I've updated comments/docs in code to reflect this.

Reply:
This type of matching seems very specific to path matching, but StringMatcher can be used for things other than paths, so I'm not sure this is really the right place for this.  I think it would make more sense to create a new matcher for this purpose.

Reply:
It can be used in any context where simplified wildcard matching based on a delimiter is needed.

Reply:
I am ok with expanding the / to be generalized

Reply:
I think that even with making this more generalized, it still seems to be a fairly special case.  I think it makes more sense for this to be a separate matcher, just like RegexMatcher is.